service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,454 @@
+            String key = entry.getKey();
+            switch (key) {
+                case NAMES_PART_KEY: {
+                    String[] namesPart = entry.getValue();
+                    return getAllByPartOfNames(namesPart);
+                }
+                case DESCRIPTION_PART_KEY: {
+                    String[] descriptionsPart = entry.getValue();
+                    return getAllByPartOfDescriptions(descriptionsPart);
+                }
+                case TAG_NAMES_KEY: {
+                    String[] tagNames = entry.getValue();
+                    return getAllByTagNames(tagNames);
+                }
+                case SORT_FIELDS_KEY: {
+                    String[] sortFields = entry.getValue();
+                    String[] orders = reqParams.get(ORDER_KEY);
+                    if (orders == null ||
+                            (!orders[0].equalsIgnoreCase("desc") && !orders[0].equalsIgnoreCase("asc"))) {
+                        throw new IllegalRequestParameterException("There is no correct order-parameter in request");
+                    }
+                    return getAllSorted(sortFields, orders[0]);
+                }
+            }
+        }
+
+        return getAll();
+    }
+
+
+    /**
+     * This method gets GiftCertificate entity from dao layer with given id and converts it to GiftCertificateDto.
+     * The method also calls {@link #fillCertificateDtoWithTags(GiftCertificateDto)} for filling gotten GiftCertificateDto
+     * with list of TagDto.
+     *
+     * @param id id of necessary entity.
+     * @return GiftCertificateDto with id and tags.
+     * @throws GiftCertificateNotFoundException if there is no entity with given id in database.
+     * @since 1.0
+     */
+    @Override
+    public GiftCertificateDto getById(long id) {
+        Optional<GiftCertificate> certificateOpt = giftCertificateDao.getById(id);
+
+        GiftCertificate giftCertificate = certificateOpt.orElseThrow(() ->
+                new GiftCertificateNotFoundException(String.format("Can't find a certificate with id: %d", id)));
+
+        GiftCertificateDto giftCertificateDto = modelMapper.map(giftCertificate, GiftCertificateDto.class);
+        fillCertificateDtoWithTags(giftCertificateDto);
+        return giftCertificateDto;
+    }
+
+    /**
+     * This method set date fields for GiftCertificate entity and saves it in database. Also the method link passed tags with the GiftCertificate
+     * entity and, if the tag doesn't exist in db, creates it.
+     *
+     * @param certificate GiftCertificate entity without date fields and id for saving.
+     * @return GiftCertificateDto with id and tags.
+     * @since 1.0
+     */
+    @Override
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public GiftCertificateDto save(GiftCertificateDto certificate) {
+        GiftCertificate giftCertificate = modelMapper.map(certificate, GiftCertificate.class);
```

Comment:
So, Do you have some validation like unique name?

Reply:
commit 23ff84f